terraform/interpolate.go

```diff
@@ -566,49 +572,70 @@ func (i *Interpolater) interpolateComplexTypeAttribute(
-	otherListKey := regexp.MustCompile("^" + resourceID + "\\.([^#]+)$")
-	for id, _ := range attributes {
-		if numberedListKey.MatchString(id) {
-			numericKeys = append(numericKeys, id)
+	// We can now distinguish between lists and maps in state by the count field:
+	//    - lists (and by extension, sets) use the traditional .# notation
+	//    - maps use the newer .% notation
+	// Consequently here we can decide how to deal with the keys appropriately
+	// based on whether the type is a map of list.
+	if lengthAttr, isList := attributes[resourceID+".#"]; isList {
+		log.Printf("[DEBUG] Interpolating computed list element attribute %s (%s)",
+			resourceID, lengthAttr)
+
+		// In Terraform's internal dotted representation of list-like attributes, the
+		// ".#" count field is marked as unknown to indicate "this whole list is
+		// unknown". We must honor that meaning here so computed references can be
+		// treated properly during the plan phase.
+		if lengthAttr == config.UnknownVariableValue {
+			return unknownVariable(), nil
 		}
-		if submatches := otherListKey.FindAllStringSubmatch(id, -1); len(submatches) > 0 {
-			allKeys = append(allKeys, submatches[0][1])
+
+		var keys []string
+		listElementKey := regexp.MustCompile("^" + resourceID + "\\.[0-9]+$")
+		for id, _ := range attributes {
+			if listElementKey.MatchString(id) {
+				keys = append(keys, id)
+			}
 		}
-	}
 
-	if len(numericKeys) == len(allKeys) {
-		// This is a list
 		var members []string
-		for _, key := range numericKeys {
+		for _, key := range keys {
 			members = append(members, attributes[key])
 		}
+		// This behaviour still seems very broken to me... it retains BC but is
+		// probably going to cause problems in future
 		sort.Strings(members)
+
 		return hil.InterfaceToVariable(members)
-	} else {
-		// This is a map
+	}
+
+	if lengthAttr, isMap := attributes[resourceID+".%"]; isMap {
+		log.Printf("[DEBUG] Interpolating computed map element attribute %s (%s)",
+			resourceID, lengthAttr)
+
+		// In Terraform's internal dotted representation of map attributes, the
+		// ".%" count field is marked as unknown to indicate "this whole list is
+		// unknown". We must honor that meaning here so computed references can be
+		// treated properly during the plan phase.
+		if lengthAttr == config.UnknownVariableValue {
+			return unknownVariable(), nil
+		}
+
+		var keys []string
+		mapElementKey := regexp.MustCompile("^" + resourceID + "\\.([^%]+)$")
+		for id, _ := range attributes {
+			if submatches := mapElementKey.FindAllStringSubmatch(id, -1); len(submatches) > 0 {
+				keys = append(keys, submatches[0][1])
+			}
+		}
+
 		members := make(map[string]interface{})
-		for _, key := range allKeys {
+		for _, key := range keys {
 			members[key] = attributes[resourceID+"."+key]
 		}
 		return hil.InterfaceToVariable(members)
 	}
+
+	return ast.Variable{}, fmt.Errorf("No complex type %s found", resourceID)
```

Comment:
I wonder if the solution to #2143 is hiding somewhere in an unnecessarily greedy regexp matcher in here, or if it's a `helper/schema` bug instead.

Non-critical color commentary shall continue! 😄